Add human readable breadcrumbs for all remaining management routes

FrankeLukas opened the pull request:
### Checklist

- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context

Not all routes were displaying human readable breadcrumbs yet.

### Description

This changes the approach to no longer require child routes to be set up. Instead, routes are registered in the navbar, translations added in a map and IDs resolved using the respective services and not resolvers. (See the new documentation and the inline code if you are interested in the details).

### Steps for Testing

Below I listed all routes that should now display proper crumbs. It should 
display translated strings and titles instead of IDs where possible (I noted the exceptions). All breadcrumbs should be clickable (and take you somewhere instead of erroring).

I have implemented workarounds for routes which do not correspond to our ideal layout. Those will be refactored in a follow-up PR (to not make this more complex than it already is).

```
Server Administration:
- Upcoming Exams
- User Management
  - Create
  - Single user view (displays user login instead of ID)
  - Single user edit (displays user login instead of ID)
- System Notifications
  - Create
  - Single notification view (displays ID as there is no title)
  - Single notification edit (displays ID as there is no title)
- Featurer toggles
- User statistics
- Metrics
- Health
- Configuration
- Audits
- Logs
- API

Course Management:
- Create new course
- Single course
  - Edit
  - Q&A
  - All three course groups
  - Learning goal management
    - Create
    - Edit
  - Assessment Dashboard
  - Instructor Dashboard
  - Course Scores

Lectures (part of course management):
- Create
- View
- Edit
- Attachments
- Units
  - Create: Text, Video, Attachment, Exercise (do not display an ID as there is not a single view outside of the container)
  - Edit: Text, Video, Attachment (do not display an ID as there is not a single view outside of the container)

Exercises (part of course management):
- For apollon diagram/drag and drop quiz:
  - View/Open, Create
- For all other exercise types:
  - Create, Edit, Scores, Participations
- Quiz exercises only: Preview, Sample Solution, Statistics, Edit, Re-evaluate
- For programming, modeling, text, and file upload exercises: View, Submissions and Assessments/Grading, Plagiarism Detection
- For text/modeling exercises: Creating example submissions and solutions (do not display IDs or new but just "Example Solution"/"Example Submissions Editor" instead)
- Exercise Assessment Dashboard for all exercises
- Exercise Instructor Dashboard for all exercises
- Assess exercises (all types)
- For programming exercises:
  - Code editor
  - Hints: View, Create, Edit

Exams (part of course management):
- Create
- For a single exam:
  - View/Id or checklist (displays exam title)
  - Edit
  - Student Exams
  - Assessment Dashboard
  - Students
  - Test runs:
    - Start (conduction) displays "Exam"
    - Assessment dashboard
    - Summary
  - Scores
  - Exercise groups:
    - Exercise buttons work the same as for course exercises
    - "Add" buttons don't have extra breadcrumbs (only display until "exercise groups")
```

The dropdowns in the navigation bar should also all still work.

### Test Coverage

app/shared/layouts/navbar/navbar.component.ts: `0%` -> `81.01%`

airelawaleria reviewed (approved):
Tested in the testing session. Basically everything works as described. However, we noticed something, most probably not yet implmented in this pr, that whenever one clicks on the Exercise breadcrumb (for a particular exercise) it never works (one can only go to the overview over all exercises, but not to that particular one). Also, when navigating to an assessment dashboard, the previously visited pages are not displayed correctly in the breadcrumbs (e.g. going to the assessment dashboard from the exam page, displayes as preceeding steps the overview over all exercises in the course. 

We assume it wasn't intended to fix in this pr, therefore approve

simonzilker reviewed (approved):
what @airelawaleria

marvinchimeka reviewed (approved):
Tested in testing session. Confirm what @airelawaleria

JohannesWeiss reviewed (dismissed):
Tested in testing session. Nice addition!
Aside from the aforementioned small issue, everything worked.

tiffca reviewed (approved):
most breadcrumbs worked, hence approving

Noticed the following:
- not sure what the code editor breadcrumb is supposed to look like, it still shows an ID 
![image](https://user-images.githubusercontent.com/63906424/108041879-6c783580-703f-11eb-953d-3cd6a4bffe73.png)

In the regular exercise assessment dashboard, clicking on "tiff 2873" nor "assessment dashboard" doesn't work.
![image](https://user-images.githubusercontent.com/63906424/108042036-9598c600-703f-11eb-9866-ebd19a73fb9a.png)

for the exams, clicking on an exercise group brings us to a breadcrumb that doesn't seem right, it looks similar to the regular exercise breadcrumb, hence there's no way to go back (you might have mentioned this isn't implemented yet)
![image](https://user-images.githubusercontent.com/63906424/108042201-caa51880-703f-11eb-9196-ca4a14224a02.png)

valeriebucher reviewed (approved):
Looks good but can confirm @tiffca 's comment

EneaGore reviewed (approved):
With exception of what has been already mentioned looks good on ts1.

DenizShabani reviewed (dismissed):
The same goes for me as well.

FrankeLukas commented:
> whenever one clicks on the Exercise breadcrumb (for a particular exercise) it never works (one can only go to the overview over all exercises, but not to that particular one).

Unfortunately there is no generic details page for exercises (i.e. when we don't know the type), so that is indeed expected until we add one.

> Also, when navigating to an assessment dashboard, the previously visited pages are not displayed correctly in the breadcrumbs (e.g. going to the assessment dashboard from the exam page, displayes as preceeding steps the overview over all exercises in the course.

> for the exams, clicking on an exercise group brings us to a breadcrumb that doesn't seem right, it looks similar to the regular exercise breadcrumb, hence there's no way to go back (you might have mentioned this isn't implemented yet). for example, clicking on 'exercises' brings us back to regular exercises, not the exam exercise groups

Breadcrumbs display depending on where in the page tree we are, not depending on the history, which is unfortunate in those cases. We might want to think about special casing here in the future. (Or adding some extra buttons.)

> In the regular exercise assessment dashboard, clicking on "tiff 2873" nor "assessment dashboard" doesn't work.

"tiff 2873" not working is the same problem as the first point in this post. "Assessment dashboard" only appears to not work since it just redirects you to the page you are already on. We could generally hide the last breadcrumb in the chain to avoid such confusion. (And it usually does not have advantages, at least I can't think of a scenario right now where you *need* to navigate to the page you're already on.)

> not sure what the code editor breadcrumb is supposed to look like, it still shows an ID

There does not exist a title for programming submissions afaik, so displaying an ID here is correct.

philgabr reviewed (dismissed):
Nice improvements!

What I noticed (additionally to already mentioned exceptions) is whenever clicking on a course related button, the breadcrumbs display the id instead of the course name for a blink before then displaying the name correctly.

FrankeLukas commented:
> What I noticed (additionally to already mentioned exceptions) is whenever clicking on a course related button, the breadcrumbs display the id instead of the course name for a blink before then displaying the name correctly.

Good observation. We indeed display the IDs until the titles are fetched from the server, so this is also expected. I will add caching in a future PR which will help mitigating this.

SimonEntholzer reviewed (dismissed):
testing locally worked fine. 
Code changes seem fine too, almost all of them are deletions anyway.

krusche reviewed (changes requested):
I have 3 changes requests:
1) we need client tests that cover the new and rather complex functionality
2) please try to split the code into smaller pieces: clean code actually requires methods to have less than 10 lines of code ;-)
3) make sure that the new code can never ever break something in some other places. Ideally we have a try catch around the whole `buildBreadcrumbs`

Please in general try to program very defensively when it comes to errors

FrankeLukas commented:
Finally updated, I added a try catch and refactored the code into smaller methods with more documentation. I also added several testcases so the coverage of the entire component went from 0% to over 80%.

tiffca reviewed (approved):
mostly everything worked for me, hence approving, but ran into the below behavior. 


Error when editing learning goal. 
Go to course management -> learning goals -> edit learning goal 

instrument.js:109 ERROR TypeError: Cannot read property 'title' of undefined
    at E.initializeForm (learning-goal-form.component.ts:110)
    at E.ngOnInit (learning-goal-form.component.ts:102)
    at yn (core.js:2521)
    at xn (core.js:2492)
    at Pn (core.js:2443)

for assessments, there is no way to get back to the assessment dashboard from an assessment. Same for exam assessment dashboard behavior, and test run assessment behavior.  

programming exercise assesment
![image](https://user-images.githubusercontent.com/63906424/109490433-81fe4e00-7a88-11eb-8aae-757b457c29a4.png)

text assessment 
![image](https://user-images.githubusercontent.com/63906424/109490453-8a568900-7a88-11eb-88e8-7dff894a828c.png)
clicking on submissions doesn't do anything

FrankeLukas commented:
@tiffca thanks for the review. Both issues are fixed now.

iozbeyli commented on src/test/javascript/spec/component/shared/navbar.component.spec.ts:
```diff
@@ -0,0 +1,494 @@
+
+            fixture.detectChanges();
+
+            expect(courseManagementStub).to.have.been.calledWith(1);
```
It might be better to change the course id and exercise id between tests to see if they are working with other values.

FrankeLukas replied:
Changing values isn't necessary imo (and wouldn't allow me to reuse as much). Either it can parse numbers or it can't.

iozbeyli commented on src/main/webapp/app/shared/layouts/navbar/navbar.component.html:
```diff
@@ -172,11 +172,13 @@
-                    </li>
+                    <div *ngIf="languages && languages.length > 1">
+                        <li *ngFor="let language of languages">
+                            <a class="dropdown-item" [jhiActiveMenu]="language" href="javascript:void(0);" (click)="changeLanguage(language); collapseNavbar()">{{
```
Why anchor if it does not go anywhere? javascript:void(0) violates Content Security Policy they say :D

FrankeLukas replied:
I didn't change that part just added a `div` with an `ngIf`.

FrankeLukas commented:
@iozbeyli I removed the `href` void tags.

airelawaleria reviewed (approved):
Since the previously mentioned issue about individual exercises it intented, we approve, because everything else was re-tested and worked as expected :)

JohannesWeiss reviewed (approved):
Tested on ts1 in testing session, works as intended with the aforementioned non-issue.

marvinchimeka reviewed (approved):
Re-Tested in testing session. 👍

JannisBittlmayer reviewed (approved):
Everything worked except one thing. For exams -> exercise group with a programming exercise, when you click on "Edit in editor" for the programming exercise, the breadcrumb shows -1 as the last part.
This doesn't seem intended but since I'm not sure I'll approve.
![grafik](https://user-images.githubusercontent.com/73840230/109673849-b819fb80-7b76-11eb-8604-ce399107323d.png)
